### DISCUSSION TITLE
SonarCloud doesn't detect tests for routes.js

### DISCUSSION THREAD

**Discussion by uo289321** [Arquisoft/faq#67] [Q&A]
Arquisoft/wiq_es05b#167
In the develop branch, after doing the tests for all the routes.js, SonarCloud doesn't link the tests with the files, and it says that there's 0% coverage, even after changing the names in case it didn't pick them up because of that. 

![image](https://github.com/Arquisoft/faq/assets/144900498/e690a595-0775-4833-b1f8-7dbf8574fad3)

In local, if you do "npm test -- --coverage", it says there's a percentage.

![image](https://github.com/Arquisoft/faq/assets/144900498/99c62fcb-d8f7-45fa-9b42-241cc4995a5e)

(Jordi is the question service)

**Comment by pglez82** [comment]
I have accessed your sonarcloud and I get different results. The ones that I get are very similar to those showed by jest:

![imagen](https://github.com/Arquisoft/faq/assets/10683040/0bec09a6-51e8-4eac-99cc-c629b8b83e37)

I am checking the main branch coverage. Maybe are you checking another branch in SonarCloud?

**Reply by uo289321** [reply]
Sorry I didn't specify, the coverage needs to be checked on the branch develop, that it's the one with the updated tests.

**Reply by algarfer** [reply]
Hi Pablo
We have just performed a merge to master with several tests and new functionality that was not fully tested.

The current [report](https://sonarcloud.io/summary/new_code?id=Arquisoft_wiq_es05b) of sonarcloud for the master branch displays 66.3% coverage for the new code and 66.7% for the overall code:
![image](https://github.com/Arquisoft/faq/assets/58784148/bf7617b4-90b5-495f-acca-8874c019b8d1)
![image](https://github.com/Arquisoft/faq/assets/58784148/9b41956b-0683-4c50-a58a-510416e42c4f)

If we have a look at the sonarcloud coverage report there are some files with low coverage like these:
![image](https://github.com/Arquisoft/faq/assets/58784148/3de0faa3-9a5c-4c7b-a33a-1f29a3fb1eb6)

And this does not meet the coverage reported by jest on the actions [log](https://github.com/Arquisoft/wiq_es05b/actions/runs/8895541866/job/24426082883) since we have approximately 83.16% (supposing the overall coverage is the mean of the tests coverage):

![image](https://github.com/Arquisoft/faq/assets/58784148/65393517-d494-43e8-8780-d799659c222e)

![image](https://github.com/Arquisoft/faq/assets/58784148/e4556d10-b1e7-49da-9679-7003789b92bf)
![image](https://github.com/Arquisoft/faq/assets/58784148/8f63cf0c-c4a0-47e4-b967-773e1f3dbede)
![image](https://github.com/Arquisoft/faq/assets/58784148/5ee1a727-3f6a-497a-b885-761955cf575d)
![image](https://github.com/Arquisoft/faq/assets/58784148/51adc5fa-8ff5-4590-bb27-f907c68832ed)
![image](https://github.com/Arquisoft/faq/assets/58784148/19f9ea63-c903-4a93-b5cd-ece07bb05810)
![image](https://github.com/Arquisoft/faq/assets/58784148/b3f69d3b-703b-43d2-a23c-37663d0cbec9)

Do you know if sonarcloud is misinterpreting the results or how we can fix this problem?

**Comment by pglez82** [comment]
Evidentemente hay algún problema, que a simple vista no estoy viendo. Donde realmente hay que mirar es en el log de la action que está corriendo los tests antes de subir el informe, pero ahí tampoco veo nada raro porque la cobertura que está generando está bien (la que os sale a vosotros en local).

Lo único que se me ocurre es subir el coverage desde local, sin pasar por actions, para intentar ver el problema que hay.

Para esto deberíais de correr en local los tests y generar en local el reporte de cobertura. A continuación, necesitareis el cliente de sonar cloud en linea de comando.

https://docs.sonarsource.com/sonarqube/latest/analyzing-source-code/scanners/sonarscanner/

Alguna vez que me ha pasado esto que os ha pasado a vosotros he acabado encontrando el fallo gracias a este cliente.

Ya me decís si os sirve para algo.

**Reply by algarfer** [reply]
He estado probando con sonar en local y creo que el problema puede estar relacionado con esto:

![image](https://github.com/Arquisoft/faq/assets/58784148/64264ab6-863d-4631-bb86-6732228aac21)

Buscando el primer error encontré esta [issue](https://community.sonarsource.com/t/found-inconsistencies-in-coverage-report/90928) en la pagina de sonar. La solución que le dan es mirar el numero de líneas del fichero correspondiente para ver si cuadra. En nuestro caso, mirando la [pestaña de code](https://sonarcloud.io/code?id=Arquisoft_wiq_es05b&selected=Arquisoft_wiq_es05b%3Agatewayservice%2Froutes), uno de los ficheros que tiene 0% de coverage es userRoutes.js, el cual tiene 122 lineas, pero en el resumen antes de entrar al archivo marca 100 lineas.

![image](https://github.com/Arquisoft/faq/assets/58784148/cd5a7616-1ee9-444b-97f4-e92418fd1ee9)
![image](https://github.com/Arquisoft/faq/assets/58784148/5d75b63d-eec1-4d83-9389-045f04d58240)

Revisando una de las issues de github que se mencionaba en el foro de sonar aparecía referenciada esta https://github.com/Arquisoft/wiq_en3b/issues/88. En su caso aparentemente se soluciono el problema metiendo unos ficheros de configuracion para jest, en mi caso tampoco se solucionó.

**Reply by pglez82** [reply]
Esto es raro:

![imagen](https://github.com/Arquisoft/faq/assets/10683040/7002c68e-f4b1-40eb-bae1-c39e36410e1c)

No se porque hay un informe de cobertura ahí, no debería de estar ni de subirse.

Por otra parte, prueba a subir módulo a módulo (en lugar de todos a la vez) a ver si no te da esos warnings. Está claro que ahí está el problema. De alguna manera hay archivos en los que no está cuadrando el informe dado por jest y luego los fuentes, por eso creo que los ignora. Si subes uno a uno quizás puedas ver que parte está dando problema. 

Lo de que no cuadre los fuentes con los informes es más típico cuando se usa typescript porque hay una traducción previa, pero no veo que lo estés usando.

Probad eso y me decís. Sino mañana le intento echar un ojo clonándome el repositorio como último recurso, a ver si veo algo.

**Reply by algarfer** [reply]
Ese informe de cobertura es de la librería [difunc](https://www.npmjs.com/package/difunc?activeTab=readme), que es una dependencia de la librería de OpenAPI. Esta ahí por que esta subida a npm con el resto del código. He probado a eliminar la carpeta pero poco ha cambiado.

Si muevo el sonar-project.properties dentro del servicio que te cite en el mensaje anterior los warings desaparecen.Por regla general el coverage de los archivos es menor que el que indica .

Si lo ejecuto desde la raiz del proyecto pero cambiando el sonar.sources para que solo mire ese servicio vuelven a aparecer los warnings, siguen apareciendo unresolved paths tanto en algunos archivos del propio servicio como de otros. Y la excepción IllegalArgumenException la lanza al procesar otro archivo.

He probado a eliminar todas las carpetas de coverage y las he vuelto a generar pero tampoco mejora.

**Reply by algarfer** [reply]
Y tampoco estamos usando Typescript así que tampoco tendría por que no creo que lo este transpilando y sea esa la causa del problema

**Reply by pglez82** [reply]
He probado a bajar el repositorio. Poniendo esto, he quitado los errores de los paths:

```
sonar.sources=users/authservice,users/userservice,gatewayservice,webapp/src,userhistory,jordi
```
Yo creo que se lia porque las rutas que hay en el lcov.info son relativas a el path en donde ejecutas los tests, que sería en este caso cada una de las carpetas generales (donde está el package.json).

Probad a poner esto a ver si cambia algo. Yo lo he probado (sin subir los resultados a sonarcloud) y ya solo quedan 2 warnings de ficheros con los paths mal (que es el que hablamos antes que está en el node modules y que habría que excluir).

En cuanto a la excepción de que no encuentra cierta linea en el código fuente, veo que me da para el archivo userRepository.js. He buscado en vuestro repositorio y veo que hay dos, en dos servicios diferentes (authService y userService). Puede ser (aunque no debería, que se esté liando). ¿Podéis probar a refactorizar y cambiar el nombre a uno de estos dos archivos?

Por último, ejecutando en local sonar-scanner --debug podéis ver muchísima información útil para depurar el problema. Podéis redireccionar toda la salida a un archivo y subirlo aquí para poder ver que pasa en caso de que todavía no funcione la cosa.

**Reply by algarfer** [reply]
Es correcto, habíamos cambiado los paths para hacer pruebas pero se nos paso volver a ponerlo como estaba, aun así ayer los tenia en local como mencionas.

Los warning por los archivos es fácil de quitar desde el workflow, yo los tengo borrados en local. Tambien es cierto que en el log menciona que hay varias inconsistencias en el reporte de cobertura:

![image](https://github.com/Arquisoft/faq/assets/58784148/24119b5c-1360-44b7-b290-c2b9849a5380)

En un rato pruebo a refactorizar los nombres de los archivos y te comento.

Y el sonar tambien lo estaba ejecutando con `sonar.verbose=true` en local. @augustocristian nos mencionó que lo dejemos activado en el workflow en https://github.com/Arquisoft/wiq_es05b/issues/167 hasta que resolvamos el problema.

**Reply by pglez82** [reply]
Ejecutarlo en local con --debug como os dije anteriormente, no hace falta que lo hagáis en el actions. Ahí se ve todo. Podéis subir la traza aquí para verla.

**Reply by algarfer** [reply]
[report.txt](https://github.com/Arquisoft/faq/files/15180372/report.txt)

Esta vez he modificado el nombre de los archivos y ha subido al 71% y parece que ya detecta bien los ficheros.

Aun asi los porcentajes no cuadran del todo con lo que reporta jest, teniendo algunos servicios entre un 1% o 2% menos hasta otros que tienen casi un 10% menos.

Aun así me esta detectando un fichero de python, el cual ya no es necesario por lo que iba a eliminarlo y volver a medirlo.

**Reply by pglez82** [reply]
Bueno ahora ya parece que va teniendo sentido la cosa. Quizás tengáis que ignorar algún fichero que no hace falta testear y subiendo un poco la cobertura de alguno por ahí que está bajo yo creo que podéis llegar al 80% sin mucho problema.

En cuanto que las diferencias de medida. Quizás no lo midan exactamente igual, no estoy seguro. Prueba a abrir el fichero en sonarcloud y ver que las lineas que te dice que no están siendo tocadas por los tests tienen sentido o no.
